# Introduction to Pandas for Working with Tabular Data

# I. Importing Necessary Packages
The following code will include the packages you'll need for this notebook. Packages are collections of code that adds additional functionality to the core Python. It's best practice to import everything you need in one place at the top of your notebook or script.

The "as pd" part of the pandas import statement below is giving the Pandas package an alias in our notebook. This way when we want to use functions from the Pandas package we can type, for example, ```pd.DataFrame()``` instead of the longer ```pandas.DataFrame()```. An alias can be anything really, but "pd" is the alias that the Pandas user community has settled on.

In [4]:
import pandas as pd

Packages that extend the core python language, such as the one we imported above usually have a website where you can find tons of helpful information. Package websites may include, for example, "Getting Started" tutorials, in-depth user guides, an API reference that documents the particulars of every single available function, and instructions on where to ask the user community questions, submit bug reports, or make software contributions. 

If you need help, package websites are one of the first places you should look. Let's take a quick look at the Pandas website **[https://pandas.pydata.org/](https://pandas.pydata.org/)**.

# II. Introduction to Pandas Data Structures

On the Pandas website, the package developers describe the project's goal: pandas "aims to be the fundamental high-level building block for doing practical, real world data analysis in Python. Additionally, it has the broader goal of becoming the most powerful and flexible open source data analysis / manipulation tool available in any language."

Pandas is a powerful tool for working with tabular data such as data stored in spreadsheets, databases, or other table-like formats. The main data structure Pandas used to hold data is called the *DataFrame*. A DataFrame is a 2-dimensional (rows and columns) structure that can store data of different types including strings, integers, floating point values, categorical data and more. DataFrames are like a spreadsheet - think of a table of data with headings and values. Each column of a Pandas DataFrame is its own data structure called a *Series*. Both data structures (DataFrame and Series) have an *index*. You can think of an index, for now, as a row or line number, but it can be anything, even text.

Below are schematics of what a Pandas DataFrame and Series look like, where the darker grey boxes would hold headers (column names) and indexes (row names/numbers), while the lighter grey boxes would hold the data values.

<table><tr>
<!-- <td> <img src="https://pandas.pydata.org/docs/_images/01_table_dataframe.svg" alt="schematic of a dataframe" width="700"/> </td>
<td> <img src="https://pandas.pydata.org/docs/_images/01_table_series.svg" alt="schematic of a series" width="200"/> </td> -->
<td> <img src="images/01_table_dataframe.svg" alt="schematic of a dataframe" width="700"/> </td>
<td> <img src="images/01_table_series.svg" alt="schematic of a series" width="200"/> </td>
</tr></table
         
(Image Source: [Pandas Docs Getting Started Tutorial](https://pandas.pydata.org/docs/getting_started/intro_tutorials/01_table_oriented.html))

## Making a DataFrame from Scratch

You can make a dataframe programmatically, as opposed to reading data from a file. For example, let's create one from lists.

In [5]:
listOfTuples = [(1,'089',32223,59730,98468,35297),
                (2,'121',27183,56159,145165,52539),
                (3,'073',27399,50075,57786,21237),
                (4,'033',25065,59734,166234,56641),
                (5,'059',23547,49620,140298,50185),
                (6,'047',23111,44550,194029,69384),
                (7,'149',22079,40404,48773,18941),
                (8,'045',21935,44494,43929,17380),
                (9,'081',21831,39049,82910,32086),
                (10,'131',21691,43728,17786,6165)]

listOfColumnNames = ["rank","county_fips","per_capita_income"
                     ,"median_household_income","population"
                     ,"num_households"]

countyData = pd.DataFrame(listOfTuples,columns=listOfColumnNames)
        
# if placed on a line by itself, you will get pretty output of the dataframe    
countyData

rank county_fips  per_capita_income  median_household_income  population  \
0     1         089              32223                    59730       98468   
1     2         121              27183                    56159      145165   
2     3         073              27399                    50075       57786   
3     4         033              25065                    59734      166234   
4     5         059              23547                    49620      140298   
5     6         047              23111                    44550      194029   
6     7         149              22079                    40404       48773   
7     8         045              21935                    44494       43929   
8     9         081              21831                    39049       82910   
9    10         131              21691                    43728       17786   

   num_households  
0           35297  
1           52539  
2           21237  
3           56641  
4           50185  
5           69384  
6           18941  
7           17380  
8           32086  
9            6165

Our data is now in a Pandas dataframe, which has 6 named columns of data (6 series) as well as a row index (the bold column without a header).

Isolating a series from a dataframe would look like this:

In [ ]:
countyData["population"]

You can see that the rendering of a series isn't pretty like that of a dataframe. But notice how a series isn't just the data values. The index remains attached and the column name is there as well. How do I know this is a Pandas series? We can use python's built-in function ```type()``` like we did in the previous notebook.

In [ ]:
type(countyData["population"])

And for good measure...

In [ ]:
type(countyData)

## Loading Tabular Data from a File into a DataFrame

Pandas makes it very easy to load an excel file or other tabular data sources like .csv files into dataframes. 

The .csv file extension is a common file format for tabular data. CSV stands for comma separated values. Inside a CSV file, you will see rows of data with commas used between the values of each data column (i.e., "comma delimited"). Generally, we use .csv files instead of excel because excel has a limit on length (1,048,576 rows). 

For example, a raw CSV file with a header row might look like this:

<pre>
Book Title,Publisher,Price
War and Peace,Vintage Classics,12.99
"Our Bodies, Ourselves",Touchstone,48.38
Putin's Playbook,"Simon & Schuster, Inc.",14.49
</pre>

Notice that the second book title listed and the third publisher name have a comma in the data value. Those values are surrounded by quotation marks to avoid Pandas interpreting the comma as a new column. This is important to do when you are creating your own csv data.

Let's begin by loading an example data file that is .csv format into a Pandas dataframe. 

The example data used in this notebook is college football bowl data. While this particular data may not be relevant to your job or research, the data exploration and cleaning techniques we'll work through do broadly apply to any tabular data you may have (in .csv, .xls(x) or even .txt formats).  


In [6]:
bowlData = pd.read_csv('data/collegefootballbowl.csv')
bowlData

id  year        date  day         winner_tie winner_rank  \
0        1  2021  12/29/2021  Wed           Oklahoma          14   
1        2  2020  12/29/2020  Tue              Texas          20   
2        3  2019  12/31/2019  Tue              Texas         NaN   
3        4  2018  12/28/2018  Fri   Washington State          12   
4        5  2017  12/28/2017  Thu    Texas Christian          13   
...    ...   ...         ...  ...                ...         ...   
1522  1523  2004  12/30/2004  Thu  Northern Illinois         NaN   
1523  1524  2003  12/30/2003  Tue       Fresno State         NaN   
1524  1525  2002  12/31/2002  Tue       Fresno State         NaN   
1525  1526  2001  12/31/2001  Mon     Michigan State         NaN   
1526  1527  2000  12/31/2000  Sun          Air Force         NaN   

      winner_points     loser_tie loser_rank  loser_points  attendance  \
0                47        Oregon         15            32     59121.0   
1                55      Colorado        NaN            23     10822.0   
2                38          Utah         12            10     60147.0   
3                28    Iowa State         25            26     60675.0   
4                39      Stanford         15            37     57653.0   
...             ...           ...        ...           ...         ...   
1522             34          Troy        NaN            21     21456.0   
1523             17          UCLA        NaN             9     20126.0   
1524             30  Georgia Tech        NaN            21     10132.0   
1525             44  Fresno State         20            35     30456.0   
1526             37  Fresno State        NaN            34     26542.0   

                                                    mvp sponsor  \
0     Oklahoma RB Kennedy Brooks, Oklahoma S Pat Fields  Valero   
1     Texas RB Bijan Robinson, Texas LB DeMarvion Ov...  Valero   
2          Texas QB Sam Ehlinger, Texas LB Joseph Ossai  Valero   
3     Washington State QB Gardner Minshew, Washingto...  Valero   
4               TCU QB Kenny Hill, TCU LB Travin Howard  Valero   
...                                                 ...     ...   
1522  RB Dewhitt Betterson (Troy), DB Lionel Hickenb...     NaN   
1523  RB Rodney Davis (Fresno State), DL Garrett McI...     NaN   
1524  RB Rodney Davis (Fresno State), DL Jason Stewa...     NaN   
1525  WR Charles Rogers (Michigan State), DL Nick My...     NaN   
1526  QB Mike Thiessen (Air Force), LB Tim Skipper (...     NaN   

                bowl_name  
0              Alamo Bowl  
1              Alamo Bowl  
2              Alamo Bowl  
3              Alamo Bowl  
4              Alamo Bowl  
...                   ...  
1522  Silicon Valley Bowl  
1523  Silicon Valley Bowl  
1524  Silicon Valley Bowl  
1525  Silicon Valley Bowl  
1526  Silicon Valley Bowl  

[1527 rows x 14 columns]

Notice what gets printed to the screen when a dataframe has many rows. We can see the first 5 and last 5 rows of data, followed by the shape (rows, columns) of the full dataframe.

Also notice some columns contain NaN values. NaN stands for "not a number" and usually represents a missing data value of type float. We'll cover more about different missing data values, how to handle them, and the missing data features Pandas offers later.

Sidebar about data management for tabular data: It's important to at minimum create a data dictionary that describes what is in your data file, even if your data file contains column names. Often there is more information (metadata) that is required by future users of the data than just the column names and data values. Metadata for the collegefootballbowl.csv can be found in the [collegefootballbowl.txt](data/collegefootballbowl.txt) file, which tells future data users the original source of the data, when the data file was created, and contains a data dictionary that describes each column of data. This type of information is super important especially when your data values have units. Even future you may forget if your data is in Celsius or Fahrenheit, or if your precipitation data had units of cm, mm, inches, or hundredths of inches! 

## Viewing the Head and Tail of a DataFrame

```.head()``` lets us view the first N rows of a dataframe 

```.tail()``` lets us view the last N rows

Using either one of these methods on a dataframe without any additional parameters will show you 5 rows. Enter an integer as a parameter to either function to view a different number of rows.



In [ ]:
# view first 10 rows
bowlData.head(10)

In [ ]:
# view last 10 rows
bowlData.tail(10)

## Getting DataFrame Information

### Shape Property: How Many Rows / Columns?

```.shape``` returns a tuple (rows, columns) and is the most concise way to see how many total rows and columns are in a dataframe. 

In [ ]:
bowlData.shape

### Dtypes Property: Understanding the Data Types of Each Column

Often you will need to know the data type of each column, ```.dtypes``` will give you that information.

Anything column that says "object" is probably storing strings. int64 and float64 are numerical data (numbers).

In [ ]:
bowlData.dtypes

How were the data types for each column determined? 

When we read the .csv file using ```pd.read_csv()``` Pandas inferred the data type of each column based on the column's data values. If all values in a column appear to be integers, Pandas will infer that the column is data type int. Sometimes, there can be mistakes in your data file though. You could have a data column, for example winner_rank, that should contain all integers or missing values but a data entry mistake in your file has added 1 or more values that are non-numeric in that column. A lot of real data is "messy" like this. In these cases where there are mixed data types in a single column, Pandas usually reads the whole column as strings and assigns a data type of "object". This is in fact the case with our data columns winner_rank and loser_rank, which we will investigate a bit more later.

### Describe() Method: Summary of Numerical Data (count, mean, std, min, quartiles, max)

You can access simple statistical information for numerical data columns using the ```.describe()``` method.

In [ ]:
bowlData.describe()

Notice that ```.describe()``` returns statistical information only for numerical data columns. 

For which year was the earliest data record in our dataframe collected? We can see the answer is the "min" of the year column, 1901.

To see all columns, numeric or not, you can add a parameter to ```.describe()``` like this:

In [ ]:
bowlData.describe(include = 'all')

You can now see 3 additional statistics that operate only on non-numeric data columns: unique, top, and freq. NaN appears wherever the data type is not appropriate for the statistic. 

Who is the most common sponsor and how many times were they a sponsor? 

Looking to the sponsor column, the "top" row indicates the most common data value is Outback Steakhouse and the "freq" row indicates that Outback Steakhouse was a sponsor 26 times.


### Info() Method: Understanding all Fields, Null Values, Dtypes, Shape, Size, etc.

In [ ]:
bowlData.info()

# III. Selecting/Slicing Data with .loc[]

The Pandas function ```.loc[row, column]``` allows us to directly access rows by index and columns by name. If you have an integer in the "row" part, Pandas assumes this is the index of the row.

## Single Cell

In [ ]:
# SELECT a single cell - the attendance column where row index is 1
bowlData.loc[1 , 'attendance']

## Single row

In [ ]:
# SELECT whole row of data where row index is 1
bowlData.loc[1 , :]

## Single Column

In [ ]:
# SELECT a single column of data, just the atttendance column
bowlData.loc[: , 'attendance']

## Slice of Rows

In [ ]:
# SELECT a "slice" of rows where index is 1 to 6
bowlData.loc[1:6 , :]

Notice that a slice of a Pandas dataframe is inclusive of the ending index 6, such that a slice of rows 1:6 returns 6 rows. Just a quick note that this is not always the case with other Python packages. Array slicing with the Numpy or Xarray packages, for example, are exclusive of the ending index, but we'll cover that in a different notebook.

## Slice Containing All Rows, but Only Certain Columns

In [ ]:
# SELECT all rows but only a  "slice" of columns 
bowlData.loc[: , 'year':'winner_rank']

## Slice Containing Certain Rows, Certain Columns

In [ ]:
# SELECT a "slice" of cells where row index is between 10 and 15 and only a few columns
bowlData.loc[10:15 , 'year':'winner_rank']

## All Rows Where Column has Certain Value

In [ ]:
# this will find all rows in the dataframe where year column is equal to 1901
bowlData.loc[bowlData['year'] == 1901]

This returned only one row. If there were more rows where the year column was equal to 1901, then there would be more rows returned.

In [ ]:
# this will find the locations in the dataset where sponsor is null (missing data)
bowlData.loc[bowlData['sponsor'].isna()]

Notice that when your output exceeds twenty lines Jupyter will format nicely and show a bit at the ending and a bit at the end. It will do the same thing if your output has too many columns.

## Rows Based on Column Comparison

If we want to return all the games where the winner was ranked below (higher number) the loser (lower number), we could do it this way.

In [ ]:
# find all rows where the winner ranked below the loser
bowlData.loc[bowlData['winner_rank'] > bowlData['loser_rank']]

Be careful in your analysis! Do you notice anything suspicious about the results that were returned?

We're looking for cases where the underdog won. For example, a team that was ranked 20th, beat the team that was ranked 10th (see index 1429). If you look closely in the results above you can see multiple examples of results returned that are NOT what we asked for. For example, index 1427 shows the winner had a better rank (8th) than the loser (13th). So, we've got incorrect results but did not receive an error message. **Why did this happen??**

This comes back to the data type of the columns winner_rank and loser_rank. A ranking should be a numerical data type (such as int or float). If these columns had a numeric data type then we shouldn't have experienced any problems using the greater than operator. But if those columns contain non-numeric data (dtype object, which usually indicates string data) then that could yield unexpected results when comparing if one string is greater than another. Let's take a look at the data type for winner_rank and loser_rank.

In [ ]:
bowlData[['winner_rank','loser_rank']].dtypes

Uh oh! We can see that Pandas determined the data type of those columns to be object, which is non-numeric. We discussed briefly already why this might happen. This should indicate to us that there must be some "messy" data somewhere in those columns, some data values that don't look like numbers. And due to this, Pandas is treating those data columns like strings instead of numbers. So, when we ask if the winner_rank is greater than the loser_rank, the process of comparing two string values with the greater than operator returns unexpected results.

When beginning work with any dataset it is best to do some data cleaning first to make sure your columns are of the expected data types and remove any whacky data values in order to avoid problems like this one. Otherwise, you might easily miss the fact that your code isn't doing what you intended.

Let's see what the problem is, clean up those two columns, ensure their dtype is numeric, and try our selection again.

We'll start by using a pandas function that we haven't seen yet: ```.unique()``` on the winner_rank column of the dataframe.

In [ ]:
# look at all the unique data values in the winner_rank column
bowlData['winner_rank'].unique()

Look at that! Someone has entered a value of Pennsylvania in the column of winner_rank somewhere in the data file. That doesn't make sense at all. Notice, we also have some missing data (nan), but that shouldn't cause us problems here.

Let's force the winner_rank data column to be numeric using another pandas function that we haven't seen yet: ```pd.to_numeric()```

In [7]:
# reassign all the values in the column winner_rank with numeric data values
# any value that does not look like a number will be changed to the missing data value
bowlData['winner_rank'] = pd.to_numeric(bowlData['winner_rank'],errors='coerce')

# look again at the unique data values
bowlData['winner_rank'].unique()

array([14., 20., nan, 12., 13., 11., 10., 15., 16., 25., 18., 24., 22.,
        9., 17., 21., 19., 23.,  8.,  6.,  4.,  2.,  3.,  1.,  5.,  7.])

In [ ]:
# look at the new data type
bowlData['winner_rank'].dtype

How did we know what parameters to enter in ```pd.to_numeric()```? That information can be found in the Pandas documentation on the [API reference page for pd.to_numeric()](https://pandas.pydata.org/docs/reference/api/pandas.to_numeric.html). Or you could have done a quick Google search for something like "how to use pandas to_numeric", which will bring up a great AI generated answer (at least it does in the USA), the link to the Pandas API reference page, and many other websites that demonstrate how to use the function.

Let's check out what the problem is with loser_rank.

In [ ]:
# look at all the unique data values
bowlData['loser_rank'].unique()

Again, there's a 'TN' in the data values that doesn't make any sense. We'll do the same process to clean the loser_rank column and convert to a numeric data type.

In [8]:
# reassign all the values in the column with numeric data values
# any value that does not look like a number will be changed to the missing data value
bowlData['loser_rank'] = pd.to_numeric(bowlData['loser_rank'],errors='coerce')

# look again at the unique data values
bowlData['loser_rank'].unique()

array([15., nan, 12., 25., 11., 22., 20., 14., 18.,  4., 24., 19., 17.,
       13., 16., 23., 21.,  7.,  9., 10.,  6.,  1.,  3.,  8.,  2.,  5.])

In [ ]:
# look at the new data type
bowlData['loser_rank'].dtype

Lastly, let's try the row selection by column comparison again.

In [ ]:
# finds all rows where the winner ranked below the loser
bowlData.loc[bowlData['winner_rank'] > bowlData['loser_rank']]

Notice how the result now returns 215 rows whereas before we got 255 rows. Our original selection returned 40 incorrect results!

# IV. Selecting data with .query()

Similar to ```.loc[]```, the ```.query()``` method can be used to select rows in a dataframe. Inside the ```.query()``` we can put a query that looks a little bit like a database query. You may like this method if you have experience working with databases.

## Rows Where Column has Certain Numerical Value

In [ ]:
# find all rows where year is equal to 1901 using .query
bowlData.query("year == 1901")

## Rows Where Column has String Value

In [ ]:
# find all rows where the bowl_name is "Rose Bowl"
# notice the single quotes around the string "Rose Bowl"
bowlData.query("bowl_name == 'Rose Bowl'")

## Rows Based on Substring Comparison

In [ ]:
# Grab all rows where the word "State" appears in the winner_tie column
bowlData.query("winner_tie.str.contains('State')")

## Rows Based on Column Comparison

In [ ]:
# find all rows where the loser ranked higher (smaller number) than the winner (larger number) using .query
bowlData.query("loser_rank < winner_rank")


## Rows Based on Comparison with a Variable

In [ ]:
# First, let's get the mean winner_points x 2
# .mean() is calculating the mean of the entire winner_points column
twiceTheMean = bowlData.winner_points.mean() * 2

# then, we can use that value to query for winners with more than twice the mean
bowlData.query("winner_points > @twiceTheMean")

## Multiple Criteria Query

In [ ]:
# Find all the rows where the winner is Alabama, 
# AND Alabama had more than 50 points, 
# AND they weren't ranked number 1.
bowlData.query("(winner_tie == 'Alabama') and (winner_points > 50) and (winner_rank > 1)")

# V. Querying a Dataframe without .query()

Different syntax can be used to accomplish the same data queries we covered above without using ```.query()```. This syntax may be preferable if you don't have experience working with databases.

In [ ]:
# find all rows where year is equal to 1901 using .query

# copy of code from the .query() section
# bowlData.query("year == 1901") 

# alternative query syntax
bowlData[bowlData.year==1901]

In [ ]:
# find all rows where the bowl_name is "Rose Bowl"

# copy of code from the .query section
# bowlData.query("bowl_name == 'Rose Bowl'") 

# alternative query syntax
bowlData[bowlData.bowl_name=="Rose Bowl"]

In [ ]:
# Grab all rows where the word "State" appears in the winner_tie column

# copy of code from the .query section
# bowlData.query("winner_tie.str.contains('State')")

# alternative query syntax
bowlData[bowlData.winner_tie.str.contains('State')]

In [ ]:
# find all rows where the loser ranked higher (smaller number) than the winner (larger number) using .query
# copy of code from the .query section
# bowlData.query("loser_rank < winner_rank")

# alternative query syntax
bowlData[bowlData.loser_rank < bowlData.winner_rank]

In [ ]:
# Rows Based on Comparison with a Variable

# copy of code from the .query section
# First, let's get the mean winner_points x 2
# twiceTheMean = bowlData.winner_points.mean() * 2
# then, we can use that value to query for winners with more than twice the mean
# bowlData.query("winner_points > @twiceTheMean")

# alternative query syntax
twiceTheMean = bowlData.winner_points.mean() * 2  # the same
bowlData[bowlData.winner_points > twiceTheMean]

In [ ]:
# Find all the rows where the winner is Alabama, 
# AND Alabama had more than 50 points, 
# AND they weren't ranked number 1.

# copy of code from the .query section
# bowlData.query("(winner_tie == 'Alabama') and (winner_points > 50) and (winner_rank > 1)")

# alternative query syntax
bowlData[(bowlData.winner_tie == 'Alabama') & (bowlData.winner_points > 50) & (bowlData.winner_rank > 1)]

Notice the difference here where we're using '&' to link multiple conditions together instead of how in the previous notebook Python Language Basics we learned how to use the boolean operator 'and' to link multiple conditions. The '&' is called  'bitwise and' whereas python 'and' is called a 'logical and'. Unless using ```.query()```, Pandas requires bitwise and, bitwise or, and bitwise not, which are written as: &, |, ~. Otherwise, you will get an error. 

# VI. DataFrame Manipulation

## Summary Functions

Pandas offers a handful of summary functions that can be applied to a column or columns of a dataframe (series objects). These functions are:

- ```.sum()``` Sum values of each object. 
- ```.count()``` Count non-NA values of each object. 
- ```.median()``` Median value of each object. 
- ```.quantile([0.25,0.75])``` Quantiles of each object. 
- ```.apply(function)``` Apply function to each object. 
- ```.min()``` Minimum value in each object. 
- ```.max()``` Maximum value in each object. 
- ```.mean()``` Mean value of each object. 
- ```.var()``` Variance of each object. 
- ```.std()``` Standard deviation of each object.

We won't cover all of these, but let's try a few.

In [11]:
# the mean of winner_points
print("Winners had an average of ",bowlData['winner_points'].mean())

# the median of loser_points
print("Losers had a median of ",bowlData['winner_points'].median())

# lowest score of a winning team
print(f"The highest score of a losing team is {bowlData['loser_points'].max()}")

Winners had an average of  30.25343811394892
Losers had a median of  30.0
The highest score of a losing team is 61


Wow, 61 points and still a loss... ooof.

## Sorting Data

Let's begin with a simple sort. To sort properly, we need the data column we are using for the sort to be numeric, otherwise we will get unexpected results (probably without warning or error) as we saw earlier in the notebook when we were comparing the values of winner_rank to loser_rank. 

What if we want to sort the dataframe by year but only return a few columns of data?

Let's first double check that the year column of the dataframe is numeric

In [ ]:
bowlData.year.dtype

Excellent, we should be good to proceed.

In [ ]:
# sort specific columns by year
bowlData[['year','winner_rank','loser_rank','winner_points']].sort_values(by=['year'], ascending=True)

Now, let's do a more complex sort. Sort by winner_points to find the highest score by an upset winner (winner ranked lower than loser).

We'll use the columns winner_rank and loser_rank to accomplish this sort (which we have already converted to numeric data) as well as the winner_points column.

Let's double check the data type of the winner_points column before we sort.

In [ ]:
bowlData.winner_points.dtype

Great, another numeric data column. This column is ok to sort on.

In [ ]:
# sort by winner_points to find the highest score by an upset winner
bowlData[bowlData.winner_rank > bowlData.loser_rank].sort_values(by=['winner_points'], ascending=False)

Notice that our code returned a lot of rows but we'll find the answer to our question in the first row in the winner_points column: 70.

How would we do the same sort but return only the highest score by an upset winner as opposed to all the data rows that were returned in the code above?

Since we are sorting with ascending = False, the highest winner_points is located in the first row of returned data. In this case, the index of the first row is 727. We don't know ahead of time what the index value of the first row of results with be though. Don't worry! We can grab the first row of the sorted results using the integer row position (0) as opposed to using the row index by using ```.iloc[]```.

From the results above we expect the output of the following code to be 70

In [ ]:
# sort by the winner_points to find the highest score by an upset winner
bowlData[bowlData.winner_rank > bowlData.loser_rank].sort_values(by=['winner_points'], ascending=False).iloc[0].loc['winner_points']

Wow! What just happened? We used ```.iloc[0]``` to select only the first row of the results returned by sorting and ```.loc['winner_points']``` to return the value of the winner_points column from the results return by ```.iloc[0]```. 

Are you beginning to see the power of Pandas? We can string together many functions in a row to achieve what we're looking for.

## Grouping Data

```.groupby()``` will partition data into groups for you, which you can then operate on using functions like```.mean()```, ```.sum()```, etc. Let's start with a super simple example before we try to use ```.groupby()``` on our bowlData.

In [ ]:
# a new very simple dataframe 
# pretend we have been observing two falcons and two parrots
# and we are keeping data records on their maximum observed flight speed in miles per hour
df = pd.DataFrame({'species': ['falcon', 'falcon','parrot', 'parrot'],
                   'individual': ['f01','f02','squawky','pretty boy'],
                   'max_speed_mph': [230., 240., 35., 40.]})
df

We can use ```.groupby``` to find the average max speed of each species if we group the dataframe rows by the species column and then for each group of rows (there will be two groups: falcon rows and parrot rows) take the mean of the max_speed_mph column.

In [ ]:
df.groupby("species")['max_speed_mph'].mean()

Nice! The average max speed of all the falcons is 235 mph and the average max speed of all the parrots is 37.5 mph.

Pandas ```.groupby()``` will work similarly on our much larger bowlData dataframe.

Let's group by winner_tie (which is the name of the winning team) and then find the average numer of points the winning team scored (winner_points).

In [ ]:
avgPointsByWinners = bowlData.groupby("winner_tie")["winner_points"].mean()
avgPointsByWinners

With the bowlData, it is a little more obvious to see the default behaviour of the ```.groupby()``` with regard to sorting the result. ```.groupby()``` will sort ascending on the grouping column (here, winner_tie) when returning the result, which is why our result is sorted by winner_tie from A to Z. This is the default behavior unless you specify sort=False as a parameter in ```.groupby```. 

## Renaming Columns

To change column names, the easiest way is this:

In [ ]:
# create a dictionary where key is the old name and value is the new name
columnMap = {"mvp":"most_valuable_player", "winner_tie":"winner"}

bowlData = bowlData.rename(columns=columnMap, errors="raise")

bowlData

## Handling Missing Values

Sometimes you will need to assess how much of your data is missing to make sure any statistics you apply to your data are robust. 

Sometimes having the missing data value (NaN) in your data is beneficial. Many functions will simply ignore missing data or propagate missing data values. For example, if you were to add one data value to another and the values were NaN + 5, the result would be NaN. This is often desirable.

But in some cases, like for various machine learning techniques, you may need to ensure that there are no missing data values present in your data at all. In this case you may need to fill the missing data values with a different number or drop rows with missing data from your dataframe entirely.

Let's look at the Pandas functions that can help us with missing data.

In [ ]:
# show how many missing values are present in each column
bowlData.isna().sum()

In [ ]:
# look at all the rows where winner_rank is missing
bowlData[bowlData['winner_rank'].isna()]

In [ ]:
# replace all missing values with 0 
# notice we are making a copy of our data by saving the dataframe to a new variable here
betterBowlData = bowlData.fillna(0)

# let's see if any missing values are still present - shouldn't be
betterBowlData.isna().sum()

Let's pause for a minute to think about what we just did. We filled all NaN with 0. Does that make sense? Does it make sense to have winner_points = 0 or sponsor = 0? You can see how filling NaN with another value may end up being confusing to you later. So think carefully if you really need to replace missing data values, or if filling with another value like 0 will work for your analysis. 

Instead of filling NaN with a different value, you may need to drop full data rows if there are any missing values present. This is how you would do that:

In [ ]:
# drop all rows where at least 1 column of data is NaN
lessBowlData=bowlData.dropna(how='any')
lessBowlData

When we first loaded bowlData, we originally had 1527 rows of data and now after dropping all rows that have at least one missing value, we are left with only 269 rows of data.

If the goal was to drop all rows where all columns contain the missing data value, you could use ```.dropna(how='all')```.

## Creating New Columns Derived from Existing Columns

Pandas allows us to easily use existing columns to calculate new data and save the calculations into a new column in the dataframe. 

Here's an example. Let's define a 'blowout' as when the winning team beats the losing team by 21 or more points. The task now is to create a new column in our dataframe that indicates whether each game (row) in our dataset was a blowout. We'll fill the column with values of True or False.

In [ ]:
bowlData['blowout'] = bowlData.winner_points - bowlData.loser_points >=21
bowlData

It's that simple! There is no looping required. Pandas is smart enough to do the subtraction row by row and fill the result in the appropriate place all on its own. Writing a loop would be much slower, which is the case for many Python packages. It's most efficient to use the built-in capabilities of whatever packages you're working with. Try to avoid looping wherever you can.

## Merging DataFrames Together

Let's pretend I have additonal college football bowl data hanging out in a separate csv file. The additional data has the same 'id' information as in collegefootballbowl.csv, but not all id's (1-1527) are present. How do I join this additional data to the bowlData dataframe?

In [ ]:
# load the data
moreBowlData = pd.read_csv('data/morecollegefootballbowldata.csv')
moreBowlData

Oh wow, that's not much data! But let's merge it into bowlData anyway. The merge column will be the id column since that is the only common column between the two data files.

In [ ]:
allBowlData = pd.merge(bowlData,moreBowlData,how='outer',on='id')
allBowlData

We can see that for the 3 id's where we have tickets_sold and best_selling_concession data, those data values now appear in the merged dataframe. And everywhere else in those two columns got filled with NaN. There are tons of ways to merge dataframes. Check out the [pd.merge() API reference](https://pandas.pydata.org/docs/reference/api/pandas.merge.html) for more information.

# VII. Writing a DataFrame to a File

Pandas has a function ```.to_csv()``` for writing a dataframe to a .csv file.

Let's write our bowl data to a file.

In [ ]:
bowlData.to_csv(r'data/bowlData.csv', index = None, header=True)

Locate the file you just wrote, open it, and see what it looks like. If you open the file in JupyterLab it will look like a spreadsheet. If you open the file with any other text editor, you should see the comma separated header and data values.

# VIII. Learning More About Pandas

For more about Pandas, start on the Pandas website where you can find:

- a nice cheat sheet https://pandas.pydata.org/docs/getting_started/index.html
- a long list of community developed tutorials https://pandas.pydata.org/docs/getting_started/tutorials.html#communitytutorials
- the user guide, which contains a bunch of 10 minute learning guides as well as more in-depth guides by topic https://pandas.pydata.org/docs/user_guide/index.html

# IX. Pandas Exercise Using Weather Data

Use Pandas to read, clean, manipulate, and aggregate weather data.

## Read the data file
Read the file at data/weatherdata.csv into a Pandas dataframe and render the dataframe to the screen

In [61]:
# add your code here


Show the data type of each column. 

In [62]:
# add your code here


## Clean the data

Look at all the columns of data. Do you see any mistakes?

Replace 'LosAngeles' at index 3 with 'Los Angeles'.

In [63]:
# add your code here


What data type is the windspeed_knots column? 

Type your answer:

Why did Pandas assign that data type to windspeed_knots?

Type your answer:

Judging by the data values in the windspeed_knots column, what data type should windspeed_knots probably be?

Type your answer:

Force the windspeed_knots column to be numeric, then show that the data type of the column did in fact change.

In [64]:
# add your code here


## Create a new column of data

Create a column of boolean data called ```IsRainy``` that indicates whether there was precipitation.

In [65]:
# add your code here


## Calculate average temperature by city

Calculate the average temperature for each city and save the result as new variable called ```avgT```.

In [66]:
# add your code here


Looking at the ```df``` dataframe and the ```avgT``` results, how many data values were used to calculate the average New York temperature? 

Type your answer:

## Sort the dataframe by date, ascending

In [67]:
# add your code here


## Write the dataframe to file

Write the ```df``` data to a file called ```weatherdata_yourname.csv```, replacing yourname with your first name. Do not include the index column, but do include the column names.

In [60]:
# add your code here


Does your csv file look like this inside?

<img src="images/weatherdata.png" alt="contents of cvs file" width="400"/>

If so, congratulations! You've successfully completed this exercise.